### PR TITLE
middleware(prometheus): use summaries instead of histograms

### LINKED ISSUE #925: perf: slow (*Prometheus).WroteResponse
Takes almost 2%

image

### PR DESCRIPTION
**Pull request by mmatczuk**
It is much more efficient to use summaries than histograms. This would give information on p50, p90 and p99.

Fixes #925

### CODE REVIEW

**Comment by budziam** [comment]
Can you provide information on how much time `(*Prometheus).WroteResponse` takes after this change?

**Comment by mmatczuk** [comment]
Once we deploy we can do some new benchmarking, to compare apples to apples.
Histograms are very heavy, we disable them in general during collection, this would allow not to do that.

File: middleware/testdata/TestPrometheusWrap.golden.txt [hunk 1]
```diff
@@ -1,32 +1,15 @@
...
-test_http_request_size_bytes_bucket{code="200",method="GET",le="512000"} 4
-test_http_request_size_bytes_bucket{code="200",method="GET",le="1.048576e+06"} 4
-test_http_request_size_bytes_bucket{code="200",method="GET",le="2.62144e+06"} 4
-test_http_request_size_bytes_bucket{code="200",method="GET",le="5.24288e+06"} 4
-test_http_request_size_bytes_bucket{code="200",method="GET",le="1.048576e+07"} 4
-test_http_request_size_bytes_bucket{code="200",method="GET",le="+Inf"} 4
+# TYPE test_http_request_size_bytes summary
+test_http_request_size_bytes{code="200",method="GET",quantile="0.5"} 5171
+test_http_request_size_bytes{code="200",method="GET",quantile="0.9"} 102451
+test_http_request_size_bytes{code="200",method="GET",quantile="0.99"} 102451
```
[the remark this answers is not in the document]
Comment (mmatczuk) [reply middleware/testdata/TestPrometheusWrap.golden.txt:12]: Why?

[the remark this answers is not in the document]
Comment (Choraden) [reply middleware/testdata/TestPrometheusWrap.golden.txt:12]: It's not about summary, but rather request size itself.
While duration may outline some slower requests, size does not matter that much.

[the remark this answers is not in the document]
Comment (mmatczuk) [reply middleware/testdata/TestPrometheusWrap.golden.txt:12]: Counting the sizes requires some technical prep work (wrapping) we could get rid of.

[the remark this answers is not in the document]
Comment (Choraden) [reply middleware/testdata/TestPrometheusWrap.golden.txt:12]: I'm hesitating, but it looks like we could get rid of it.

**Review by Choraden — APPROVED** [review]
LGTM

**Pull request merged**